packages/xstate-react/test/types.test.tsx

```diff
@@ -131,4 +122,75 @@ describe('useMachine', () => {
 
     render(<YesNo />);
   });
+
+  // Example from: https://github.com/davidkpiano/xstate/discussions/1534
+  it('spawned actors should be typed correctly', () => {
```

Comment:
I have no idea why but this test doesn't compile correctly on the codesandbox I've provided: https://codesandbox.io/s/thirsty-goldberg-z471l?file=/src/App.tsx . I've also rechecked by copy-pasting this code test into the codesandbox and it suffers from the same problem. It might be depending on tsconfig settings but I've tried enabling/disabled TS strict mode in the repo and on the codesandbox and it didn't change individual results :/

Reply:
Not sure why, but I'm basing the validity of this on whether it passes in this repo, not in CodeSandbox.

Reply:
This repo is not perfect for replicating all kind of behaviors for two main reasons:
- it's not using TS strict mode and in general, might also use different TS options than what a user is using
- it's not using the same TS version

Arguably the second argument is not that important because we only tend to support the latest TS version but the first problem is real. If only problems would manifest with looser configs than what we have we could just advise using stricter settings - but it's the other way around here and I believe that supporting the strict mode should be our goal.

It seems that this is indeed strict mode's problem. I've created an isolated repository to test this: https://github.com/Andarist/xstate-1622-test/commits/master

1. First I've attempted to only use the new version of `@xstate/react` -> it didn't work
2. I've tried to provide an explicit type param to the used `assign` -> it still didn't work
3. I've added a new `xstate` version and reverted the explicit type param -> it still didn't work
4. I've finally re-added that explicit type param -> it did work

And later I've confirmed that removing the strict mode and removing that explicit type param (so the situation we are having here, in this repo) removes the type problem as well.

So in a sense - this test here doesn't validate in full the user's problem. I don't believe we should try to set up things to try validate those kinds of things here, but it's IMHO an important takeaway that we shouldn't take for granted that is we don't experience type problems in this repo then people won't experience them as well.

The really weird thing for me is that this explicit type param on `assign` has been changing the return type of the whole `createMachine` call. How this can even happen if supposedly providing any explicit type param turn off inference completely?

**Without explicit type param on assign - not working**
<img width="535" alt="Screenshot 2020-11-17 at 14 22 19" src="https://user-images.githubusercontent.com/9800850/99396827-82f8de80-28e2-11eb-9236-32eca2f2cde7.png">

**With explicit type param on assign - working**
<img width="538" alt="Screenshot 2020-11-17 at 14 22 32" src="https://user-images.githubusercontent.com/9800850/99396821-812f1b00-28e2-11eb-9250-a5bab22d8814.png">

Reply:
Note: I have a potential fix for the `assign` needing an explicit type param in the pipeline but first I need to take this through the finish line: https://github.com/davidkpiano/xstate/pull/1543 , I have no idea if it's going to help with this particular case but according to my tests it has fixed other, similar situations with `assign`